File: mathsys/dev/latex/code.py
#^
#^  HEAD
#^

#> HEAD -> DATA
from .local import types
from . import dataclasses as latex
from ..parser import dataclasses as parser


#^
#^  LATEX
#^

#> LATEX -> GENERATOR
class LaTeX:
    #~ GENERATOR -> INIT
    def __init__(self) -> None: pass
    #~ GENERATOR -> RUN
    def run(self, start: parser.Start) -> str: types.clear(); return self.start(start)
    #~ GENERATOR -> START GENERATION
    def start(self, start: parser.Start) -> str:
        return str(latex.Start(
            statements = [self.level1(statement) for statement in start.statements if self.level1(statement)]
        ))
    #~ GENERATOR -> 1 LEVEL GENERATION
    def level1(self, level1: parser.Level1) -> str:
        match level1:
            case parser.Declaration(): return self.declaration(level1)
            case parser.Definition(): return self.definition(level1)
            case parser.Annotation(): return self.annotation(level1)
            case parser.Node(): return self.node(level1)
            case parser.Equation(): return self.equation(level1)
            case parser.Comment(): return self.comment(level1)
            case parser.Use(): return self.use(level1)
    #~ GENERATOR -> 1 DECLARATION GENERATION
    def declaration(self, declaration: parser.Declaration) -> str:
        if declaration.group is not None: types[declaration.variable.representation] = types.get(declaration.variable.representation, declaration.group)
        return str(latex.Declaration(
            group = declaration.group if declaration.group is not None else "",
            variable = self.variable(declaration.variable),
            expression = self.expression(declaration.expression)
        ))
    #~ GENERATOR -> 1 DEFINITION GENERATION
    def definition(self, definition: parser.Definition) -> str:
        if definition.group is not None: types[definition.variable.representation] = types.get(definition.variable.representation, definition.group)
        return str(latex.Definition(
            group = definition.group if definition.group is not None else "",
            variable = self.variable(definition.variable),
            expression = self.expression(definition.expression)
        ))
    #~ GENERATOR -> 1 ANNOTATION GENERATION
    def annotation(self, annotation: parser.Annotation) -> str:
        if annotation.group is not None: 
            for variable in annotation.variables:
                types[variable.representation] = types.get(variable.representation, annotation.group)
        return str(latex.Annotation(
            group = annotation.group if annotation.group is not None else "",
            variables = [self.variable(variable) for variable in annotation.variables]
        ))
    #~ GENERATOR -> 1 NODE GENERATION
    def node(self, node: parser.Node) -> str:
        return str(latex.Node(
            expression = self.expression(node.expression)
        ))
    #~ GENERATOR -> 1 EQUATION GENERATION
    def equation(self, equation: parser.Equation) -> str:
        return str(latex.Equation(
            leftexpression = self.expression(equation.leftexpression),
            rightexpression = self.expression(equation.rightexpression)
        ))
    #~ GENERATOR -> 1 COMMENT GENERATION
    def comment(self, comment: parser.Comment) -> str:
        return str(latex.Comment(
            text = comment.text
        ))
    #~ GENERATOR -> 1 USE GENERATION
    def use(self, use: parser.Use) -> str:
        return str(latex.Use(
            name = use.name,
            start = use.start is not None
        ))
    #~ GENERATOR -> 2 LEVEL GENERATION
    def level2(self, level2: parser.Level2) -> str:
        match level2:
            case parser.Expression(): return self.expression(level2)
    #~ GENERATOR -> 2 EXPRESSION GENERATION
    def expression(self, expression: parser.Expression) -> str:
        return str(latex.Expression(
            signs = [sign if sign is not None else "" for sign in expression.signs],
            terms = [self.level3(term) for term in expression.terms]
        ))
    #~ GENERATOR -> 3 LEVEL GENERATION
    def level3(self, level3: parser.Level3) -> str:
        match level3:
            case parser.Term(): return self.term(level3)
    #~ GENERATOR -> 3 TERM GENERATION
    def term(self, term: parser.Term) -> str:
        numerator = []
        for index in range(len(term.numerator)):
            value = self.level4(term.numerator[index])
            if index != 0:
                if isinstance(term.numerator[index - 1], parser.Factor):
                    if isinstance(term.numerator[index - 1].value, parser.Number):
                        if isinstance(term.numerator[index], parser.Factor):
                            if isinstance(term.numerator[index].value, parser.Number | parser.Infinite):
                                value = r"\cdot " + value
                    else: value = r"\cdot " + value
                else: value = r"\cdot " + value
            numerator.append(value)
        denominator = []
        for index in range(len(term.denominator)):
            value = self.level4(term.denominator[index])
            if index != 0:
                if isinstance(term.denominator[index - 1], parser.Factor):
                    if isinstance(term.denominator[index - 1].value, parser.Number):
                        if isinstance(term.denominator[index], parser.Factor):
                            if isinstance(term.denominator[index].value, parser.Number | parser.Infinite):
                                value = r"\cdot " + value
                    else: value = r"\cdot " + value
                else: value = r"\cdot " + value
            denominator.append(value)
        return str(latex.Term(
            numerator = numerator,
            denominator = denominator
        ))
    #~ GENERATOR -> 4 LEVEL GENERATION
    def level4(self, level4: parser.Level4) -> str:
        match level4:
            case parser.Factor(): return self.factor(level4)
            case parser.Limit(): return self.limit(level4)
    #~ GENERATOR -> 4 FACTOR GENERATION
    def factor(self, factor: parser.Factor) -> str:
        return str(latex.Factor(
            value = self.level5(factor.value),
            exponent = self.expression(factor.exponent) if factor.exponent is not None else None
        ))
    #~ GENERATOR -> 4 LIMIT GENERATION
    def limit(self, limit: parser.Limit) -> str:
        return str(latex.Limit(
            variable = self.variable(limit.variable),
            approach = self.expression(limit.approach),
            direction = limit.direction,
            nest = self.nest(limit.nest),
            exponent = self.expression(limit.exponent) if limit.exponent is not None else None
        ))
    #~ GENERATOR -> 5 LEVEL GENERATION
    def level5(self, level5: parser.Level5) -> str:
        match level5:
            case parser.Infinite(): return self.infinite(level5)
            case parser.Variable(): return self.variable(level5)
            case parser.Nest(): return self.nest(level5)
            case parser.Tensor(): return self.tensor(level5)
            case parser.Number(): return self.number(level5)
    #~ GENERATOR -> 5 INFINITE GENERATION
    def infinite(self, infinite: parser.Infinite) -> str: 
        return str(latex.Infinite())
    #~ GENERATOR -> 5 VARIABLE GENERATION
    def variable(self, variable: parser.Variable) -> str:
        return str(latex.Variable(
            representation = variable.representation
        ))
    #~ GENERATOR -> 5 NEST GENERATION
    def nest(self, nest: parser.Nest) -> str:
        return str(latex.Nest(
            expression = self.expression(nest.expression) if nest.expression is not None else ""
        ))
    #~ GENERATOR -> 5 TENSOR GENERATION
    def tensor(self, vector: parser.Tensor) -> str:
        return str(latex.Tensor(
            values = [self.expression(value) for value in vector.values]
        ))
    #~ GENERATOR -> 5 NUMBER GENERATION
    def number(self, number: parser.Number) -> str:
        return str(latex.Number(
            value = number.value,
            shift = number.shift
        ))

File: mathsys/dev/latex/code.ts
//^
//^ HEAD
//^

//> HEAD -> DATA
import {types} from "./local.js";
import * as latex from "./dataclasses.js";
import * as parser from "../parser/dataclasses.js";


//^
//^ LATEX
//^

//> LATEX -> GENERATOR
export class LaTeX {
    //~ GENERATOR -> CONSTRUCTOR
    constructor() {}
    //~ GENERATOR -> RUN
    run(start: parser.Start): string {
        for (const key in types) {
            delete types[key];
        }
        return this.start(start);
    }
    //~ GENERATOR -> START GENERATION
    start(start: parser.Start): string {
        return String(new latex.Start(
            start.statements.map(statement => this.level1(statement)).filter(statement => Boolean(statement))
        ));
    }
    //~ GENERATOR -> 1 LEVEL GENERATION
    level1(level1: parser.Level1): string {
        if (level1 instanceof parser.Declaration) return this.declaration(level1);
        if (level1 instanceof parser.Definition) return this.definition(level1);
        if (level1 instanceof parser.Annotation) return this.annotation(level1);
        if (level1 instanceof parser.Node) return this.node(level1);
        if (level1 instanceof parser.Equation) return this.equation(level1);
        if (level1 instanceof parser.Comment) return this.comment(level1);
        if (level1 instanceof parser.Use) return this.use(level1);
        return "";
    }
    //~ GENERATOR -> 1 DECLARATION GENERATION
    declaration(declaration: parser.Declaration): string {
        if (declaration.group !== null) {
            types[declaration.variable.representation] = declaration.variable.representation in types ? types[declaration.variable.representation] : declaration.group
        }
        return String(new latex.Declaration(
            declaration.group !== null ? declaration.group : "",
            this.variable(declaration.variable),
            this.expression(declaration.expression)
        ));
    }
    //~ GENERATOR -> 1 DEFINITION GENERATION
    definition(definition: parser.Definition): string {
        if (definition.group !== null) {
            types[definition.variable.representation] = definition.variable.representation in types ? types[definition.variable.representation] : definition.group
        }
        return String(new latex.Definition(
            definition.group !== null ? definition.group : "",
            this.variable(definition.variable),
            this.expression(definition.expression)
        ));
    }
    //~ GENERATOR -> 1 ANNOTATION GENERATION
    annotation(annotation: parser.Annotation): string {
        for (const variable of annotation.variables) {
            types[variable.representation] = variable.representation in types ? types[variable.representation] : annotation.group
        }
        return String(new latex.Annotation(
            annotation.group,
            annotation.variables.map(variable => this.variable(variable))
        ));
    }
    //~ GENERATOR -> 1 NODE GENERATION
    node(node: parser.Node): string {
        return String(new latex.Node(
            this.expression(node.expression)
        ));
    }
    //~ GENERATOR -> 1 EQUATION GENERATION
    equation(equation: parser.Equation): string {
        return String(new latex.Equation(
            this.expression(equation.leftexpression),
            this.expression(equation.rightexpression)
        ));
    }
    //~ GENERATOR -> 1 COMMENT GENERATION
    comment(comment: parser.Comment): string {
        return String(new latex.Comment(
            comment.text
        ));
    }
    //~ GENERATOR -> 1 USE GENERATION
    use(use: parser.Use): string {
        if (use.start !== null) {this.start(use.start)}
        return String(new latex.Use(
            use.name,
            use.start !== null
        ));
    }
    //~ GENERATOR -> 2 LEVEL GENERATION
    level2(level2: parser.Level2): string {
        if (level2 instanceof parser.Expression) return this.expression(level2);
        return "";
    }
    //~ GENERATOR -> 2 EXPRESSION GENERATION
    expression(expression: parser.Expression): string {
    if (!expression) {
        console.error("LaTeX.expression called with:", expression);
        throw new Error("Invalid Expression passed to LaTeX.expression");
    }
        return String(new latex.Expression(
            expression.signs.map(sign => sign !== null ? sign : ""),
            expression.terms.map(term => this.level3(term))
        ));
    }
    //~ GENERATOR -> 3 LEVEL GENERATION
    level3(level3: parser.Level3): string {
        if (level3 instanceof parser.Term) return this.term(level3);
        return "";
    }
    //~ GENERATOR -> 3 TERM GENERATION
    term(term: parser.Term): string {
        const numerator: string[] = [];
        term.numerator.forEach((item: parser.Level4, index: number) => {
            let value = this.level4(item);
            if (index !== 0) {
                if (term.numerator[index - 1] instanceof parser.Factor) {
                    if ((term.numerator[index - 1] as any).value instanceof parser.Number) {
                        if (item instanceof parser.Factor) {
                            if (item.value instanceof parser.Number || item.value instanceof parser.Infinite) {
                                value = String.raw`\cdot ` + value;
                            }
                        }
                    } else {value = String.raw`\cdot ` + value}
                } else {value = String.raw`\cdot ` + value}
            }
            numerator.push(value);
        });
        const denominator: string[] = [];
        term.denominator.forEach((item: parser.Level4, index: number) => {
            let value = this.level4(item);
            if (index !== 0) {
                if (term.denominator[index - 1] instanceof parser.Factor) {
                    if ((term.denominator[index - 1] as any).value instanceof parser.Number) {
                        if (item instanceof parser.Factor) {
                            if (item.value instanceof parser.Number || item.value instanceof parser.Infinite) {
                                value = String.raw`\cdot ` + value;
                            }
                        }
                    } else {value = String.raw`\cdot ` + value}
                } else {value = String.raw`\cdot ` + value}
            }
            denominator.push(value);
        });
        return String(new latex.Term(
            numerator,
            denominator
        ));
    }
    //~ GENERATOR -> 4 LEVEL GENERATION
    level4(level4: parser.Level4): string {
        if (level4 instanceof parser.Factor) return this.factor(level4);
        if (level4 instanceof parser.Limit) return this.limit(level4);
        return "";
    }
    //~ GENERATOR -> 4 FACTOR GENERATION
    factor(factor: parser.Factor): string {
        return String(new latex.Factor(
            this.level5(factor.value),
            factor.exponent !== null ? this.expression(factor.exponent) : null
        ));
    }
    //~ GENERATOR -> 4 LIMIT GENERATION
    limit(limit: parser.Limit): string {
        return String(new latex.Limit(
            this.variable(limit.variable),
            this.expression(limit.approach),
            limit.direction,
            this.nest(limit.nest),
            limit.exponent !== null ? this.expression(limit.exponent) : null
        ));
    }
    //~ GENERATOR -> 5 LEVEL GENERATION
    level5(level5: parser.Level5): string {
        if (level5 instanceof parser.Infinite) return this.infinite(level5);
        if (level5 instanceof parser.Variable) return this.variable(level5);
        if (level5 instanceof parser.Nest) return this.nest(level5);
        if (level5 instanceof parser.Tensor) return this.tensor(level5);
        if (level5 instanceof parser.Number) return this.number(level5);
        return "";
    }
    //~ GENERATOR -> 5 INFINITE GENERATION
    infinite(infinite: parser.Infinite): string {
        return String(new latex.Infinite());
    }
    //~ GENERATOR -> 5 VARIABLE GENERATION
    variable(variable: parser.Variable): string {
        return String(new latex.Variable(
            variable.representation
        ));
    }
    //~ GENERATOR -> 5 NEST GENERATION
    nest(nest: parser.Nest): string {
        return String(new latex.Nest(
            nest.expression !== null ? this.expression(nest.expression) : ""
        ));
    }
    //~ GENERATOR -> 5 TENSOR GENERATION
    tensor(tensor: parser.Tensor): string {
        return String(new latex.Tensor(
            tensor.values.map(value => this.expression(value))
        ));
    }
    //~ GENERATOR -> 5 NUMBER GENERATION
    number(number: parser.Number): string {
        return String(new latex.Number(
            number.value,
            number.shift
        ));
    }
}